skxray/core/accumulators/correlation/correlation.py

```diff
@@ -0,0 +1,269 @@
+# POSSIBILITY OF SUCH DAMAGE.                                          #
+########################################################################
+from __future__ import absolute_import, division, print_function
+
+from ...utils import multi_tau_lags
+from ...roi import extract_label_indices
+from .pyprocess import pyprocess
+from collections import namedtuple
+import numpy as np
+
+results = namedtuple(
+    'correlation_results',
+    ['g2', 'lag_steps', 'internal_state']
+)
+
+_internal_state = namedtuple(
+    'correlation_state',
+    ['buf',
+     'G',
+     'past_intensity',
+     'future_intensity',
+     'img_per_level',
+     'label_mask',
+     'track_level',
+     'cur',
+     'pixel_list',
+     'label_mapping',
+     ]
+)
+
+
+def _init_state(num_levels, num_bufs, labels):
+    """Initialize a stateful namedtuple for the generator-based multi-tau
+
+    Parameters
+    ----------
+    num_levels : int
+    num_bufs : int
+    labels : array
+        Two dimensional labeled array that contains ROI information
+
+    Returns
+    -------
+    internal_state : namedtuple
+        The namedtuple that contains all the state information that
+        `lazy_multi_tau` requires so that it can be used to pick up processing
+        after it was interrupted
+    """
+    label_mask, pixel_list = extract_label_indices(labels)
+    # map the indices onto a sequential list of integers starting at 1
+    label_mapping = {label: n for n, label in enumerate(
+            np.unique(label_mask))}
+    # remap the label mask to go from 0 -> max(_labels)
+    for label, n in label_mapping.items():
+        label_mask[label_mask == label] = n
+
+    # G holds the un normalized auto- correlation result. We
+    # accumulate computations into G as the algorithm proceeds.
+    G = np.zeros(((num_levels + 1) * num_bufs / 2, len(label_mapping)),
+                 dtype=np.float64)
+    # matrix for normalizing G into g2
+    past_intensity = np.zeros_like(G)
+    # matrix for normalizing G into g2
+    future_intensity = np.zeros_like(G)
+    # Ring buffer, a buffer with periodic boundary conditions.
+    # Images must be keep for up to maximum delay in buf.
+    buf = np.zeros((num_levels, num_bufs, len(pixel_list)),
+                   dtype=np.float64)
+    # to track how many images processed in each level
+    img_per_level = np.zeros(num_levels, dtype=np.int64)
+    # to track which levels have already been processed
+    track_level = np.zeros(num_levels, dtype=bool)
+    # to increment buffer
+    cur = np.ones(num_levels, dtype=np.int64)
+
+    return _internal_state(
+        buf,
+        G,
+        past_intensity,
+        future_intensity,
+        img_per_level,
+        label_mask,
+        track_level,
+        cur,
+        pixel_list,
+        label_mapping,
+    )
+
+
+def lazy_multi_tau(image_iterable, num_levels, num_bufs, labels,
+                   processing_func=None, _state=None):
```

Comment:
I do not remember if I initially suggested it, but I am not a fan of the `_` arg.